machine/v1alpha1/provider_types.go

```diff
@@ -0,0 +1,207 @@
+	// Number of desired pods. This is a pointer to distinguish between explicit zero and not specified. Defaults to 1.
+	// +optional
+	// +kubebuilder:validation:Minimum=0
+	Replicas *int `json:"replicas,omitempty"`
+
+	// NodeSelector is a selector which must be true for the pod to fit on a node.
+	// Selector which must match a node's labels for the pod to be scheduled on that node.
+	// More info: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+
+	// If specified, the pod's tolerations.
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+
+	// If specified, the pod's scheduling constraints
+	// +optional
+	Affinity *corev1.Affinity `json:"affinity,omitempty"`
+
+	// List of containers specified in the Deployment
+	// +optional
+	Containers []ContainerSpec `json:"containers"`
+}
+
+// ContainerSpec defines the properties available to override for each
+// container in a provider deployment such as Image and Args to the container’s
+// entrypoint.
+type ContainerSpec struct {
+	// Name of the container. Cannot be updated.
+	Name string `json:"name"`
+
+	// Container Image Name
+	// +optional
+	Image *ImageMeta `json:"image,omitempty"`
+
+	// Args represents extra provider specific flags that are not encoded as fields in this API.
+	// Explicit controller manager properties defined in the `Provider.ManagerSpec`
+	// will have higher precedence than those defined in `ContainerSpec.Args`.
+	// For example, `ManagerSpec.SyncPeriod` will be used instead of the
+	// container arg `--sync-period` if both are defined.
+	// The same holds for `ManagerSpec.FeatureGates` and `--feature-gates`.
+	// +optional
+	Args map[string]string `json:"args,omitempty"`
+
+	// List of environment variables to set in the container.
+	// +optional
+	Env []corev1.EnvVar `json:"env,omitempty"`
+
+	// Compute resources required by this container.
+	// +optional
+	Resources *corev1.ResourceRequirements `json:"resources,omitempty"`
+}
+
+// ImageMeta allows to customize the image used.
+type ImageMeta struct {
+	// Repository sets the container registry to pull images from.
+	// +optional
+	Repository *string `json:"repository,omitempty"`
+
+	// Name allows to specify a name for the image.
+	// +optional
+	Name *string `json:"name,omitempty"`
+
+	// Tag allows to specify a tag for the image.
+	// +optional
+	Tag *string `json:"tag,omitempty"`
+}
+
+// FetchConfiguration determines the way to fetch the components and metadata for the provider.
+type FetchConfiguration struct {
+	// URL to be used for fetching the provider’s components and metadata from a remote Github repository.
+	// For example, https://github.com/{owner}/{repository}/releases
+	// The version of the release will be `ProviderSpec.Version` if defined
+	// otherwise the `latest` version will be computed and used.
+	// +optional
+	URL *string `json:"url,omitempty"`
```

Comment:
This is a generic API, not for openshift?  We shouldn't allow this to be changed.